### PR TITLE
feature: Update copies of "user-provided"-secrets in other namespaces

### PR DESCRIPTION
**Pull request by rohitsakala**
[#174355462](https://www.pivotaltracker.com/n/projects/2192232/stories/174355462)

### CODE REVIEW

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go [hunk 1]
```diff
@@ -465,13 +466,13 @@ func (r *ReconcileQuarksSecret) createDockerConfigJSON(ctx context.Context, qsec
 // Skip creation when
 // * secret is already generated according to qsecs status field
 // * secret exists, but was not generated (user created secret)
```
Comment (edwardstudy) [inline pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:468]: Could we add comments like:

```
return existingSecret if it's not generated
```

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go [hunk 2]
```diff
@@ -500,23 +500,23 @@ func (r *ReconcileQuarksSecret) skipCreation(ctx context.Context, qsec *qsv1a1.Q
...
+		return true, existingSecret, nil
 	}
 
-	return false, nil
+	return false, nil, nil
 }
 
 // Skip copy creation when
 // * secret and qsecret (copy type) doesn't exist
 // * (q)secret exists, but it's not marked as generated (label) and a copy (annotation)
```
Comment (edwardstudy) [inline pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:511]: Same as above.

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go [hunk 3]
```diff
@@ -619,56 +627,89 @@ func (r *ReconcileQuarksSecret) createSecrets(ctx context.Context, qsec *qsv1a1.
...
+		copiedSecret.Labels = secret.Labels
+
 		if skip {
 			ctxlog.WithEvent(qsec, "SkipCreation").Infof(ctx, "Skip copy creation: Secret/QSecret '%s' must exist and have the appropriate labels and annotations to receive a copy", copy.String())
 		} else if needsCreation {
-			if err := r.createCopySecret(ctx, copiedQSec, copiedSecret, qsec.GetNamespacedName()); err != nil {
+			if err := r.createCopySecret(ctx, copiedQSec, copiedSecret, qsec.GetNamespacedName(), userCreatedSecret); err != nil {
 				return err
 			}
+			ctxlog.WithEvent(qsec, "SkipCreation").Infof(ctx, "Copy secret '%s' has been created in namespace '%s'", copy.Name, copy.Namespace)
```
Comment (edwardstudy) [inline pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:698]: => "Copied secret"

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go (hunk 1 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:468]: Done.

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go (hunk 2 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:511]: Done.

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go (hunk 3 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:698]: Done.

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go [hunk 4]
```diff
@@ -619,56 +626,89 @@ func (r *ReconcileQuarksSecret) createSecrets(ctx context.Context, qsec *qsv1a1.
...
 	if skipCreation {
 		ctxlog.WithEvent(qsec, "SkipCreation").Infof(ctx, "Skip creation: Secret '%s/%s' already exists and it's not generated", qsec.Namespace, qsec.Spec.SecretName)
+		secret = existingSecret
+
+		// Add user created annotation
+		if secret != nil {
+			if secret.Annotations == nil {
+				secret.Annotations = map[string]string{}
+			}
+			secret.Annotations[qsv1a1.AnnotationUserCreatedSecret] = "true"
```
Comment (manno) [inline pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:645]: What is this used for? I can't find code that references it.

There is also the `quarks.cloudfoundry.org/secret-kind=generated` label.

File: pkg/kube/controllers/quarkssecret/quarkssecret_controller.go [hunk 5]
```diff
@@ -97,6 +102,64 @@ func AddQuarksSecret(ctx context.Context, config *config.Config, mgr manager.Man
...
+	}
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+			secret := a.Object.(*corev1.Secret)
+
+			if skip.Reconciles(ctx, mgr.GetClient(), secret) {
+				return []reconcile.Request{}
+			}
+
+			reconciles, err := reference.GetReconciles(ctx, mgr.GetClient(), reference.ReconcileForQuarksSecret, secret, false)
```
Comment (manno) [inline pkg/kube/controllers/quarkssecret/quarkssecret_controller.go:131]: There has to be an easier way than to copy the `GetReconciles` package.

File: pkg/kube/controllers/quarkssecret/quarkssecret_controller.go [hunk 6]
```diff
@@ -97,6 +102,64 @@ func AddQuarksSecret(ctx context.Context, config *config.Config, mgr manager.Man
...
+				if err != nil {
+					ctxlog.Errorf(ctx, "could not get QuarksSecret '%s': %v", qsec.GetNamespacedName(), err)
+				}
+
+				if qsec.Status.Generated == nil {
+					qsec.Status = qsv1a1.QuarksSecretStatus{}
+				}
+				qsec.Status.Generated = pointers.Bool(false)
+				qsec.Status.LastReconcile = nil
+				err = mgr.GetClient().Status().Update(ctx, &qsec)
```
Comment (manno) [inline pkg/kube/controllers/quarkssecret/quarkssecret_controller.go:150]: This update triggers another reconcile on the qsec?

File: pkg/kube/controllers/quarkssecret/quarkssecret_controller.go [hunk 7]
```diff
@@ -130,3 +193,9 @@ func listSecrets(ctx context.Context, client crc.Client, qsec *qsv1a1.QuarksSecr
 
 	return result, nil
 }
+
+func isUserCreatedSecret(secret *corev1.Secret) bool {
+	secretLabels := secret.GetLabels()
+	_, ok := secretLabels[qsv1a1.LabelKind]
```
Comment (manno) [inline pkg/kube/controllers/quarkssecret/quarkssecret_controller.go:199]: This is the label that is set to 'generated', so the opposite?

File: pkg/kube/util/reference/reconciles.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+func GetReconcilesWithFilter(ctx context.Context, client crc.Client, reconcileType ReconcileType, object apis.Object, versionCheck bool, admitFn func(v interface{}) bool) ([]reconcile.Request, error) {
+	objReferencedBy := func(parent interface{}) (bool, error) {
+		var (
+			objectReferences map[string]bool
+			err              error
+			name             string
+		)
+
+		switch object := object.(type) {
+		case *corev1.Secret:
```
Comment (manno) [inline pkg/kube/util/reference/reconciles.go:44]: I'd get rid of the switch. This is never called for more than one type.

File: pkg/kube/util/reference/reconciles.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+func (r ReconcileType) String() string {
+	return [...]string{
+		"QuarksSecret",
+	}[r]
+}
+
+// GetReconcilesWithFilter returns reconciliation requests for the QuarksSecret
+// that reference an object. The object can be a Secret, it accepts an admit function which is used for filtering the object
+func GetReconcilesWithFilter(ctx context.Context, client crc.Client, reconcileType ReconcileType, object apis.Object, versionCheck bool, admitFn func(v interface{}) bool) ([]reconcile.Request, error) {
+	objReferencedBy := func(parent interface{}) (bool, error) {
```
Comment (manno) [inline pkg/kube/util/reference/reconciles.go:36]: The inline method is not useful.

File: pkg/kube/util/reference/reconciles.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+		_, ok := objectReferences[name]
+		return ok, nil
+	}
+
+	namespace := object.GetNamespace()
+	result := []reconcile.Request{}
+
+	log.Debugf(ctx, "Listing '%s' for '%s/%s'", reconcileType, namespace, object.GetName())
+	switch reconcileType {
+	case ReconcileForQuarksSecret:
```
Comment (manno) [inline pkg/kube/util/reference/reconciles.go:64]: Useless switch statement.

File: pkg/kube/util/reference/secret_references.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+		return getSecretRefFromQuarksSecret(object)
+	default:
+		return nil, errors.New("can't get secret references for unknown type; supported types are BOSHDeployment and QuarksStatefulSet")
+	}
+}
+
+func getSecretRefFromQuarksSecret(object qsv1a1.QuarksSecret) (map[string]bool, error) {
+	result := map[string]bool{}
+
+	result[object.Spec.SecretName] = true
```
Comment (manno) [inline pkg/kube/util/reference/secret_references.go:23]: This is a really complicated way to look at `.Spec.SecretName`.

File: pkg/kube/util/resources/fetch.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+	"context"
+
+	"github.com/pkg/errors"
+	crc "sigs.k8s.io/controller-runtime/pkg/client"
+
+	qsv1a1 "code.cloudfoundry.org/quarks-secret/pkg/kube/apis/quarkssecret/v1alpha1"
+)
+
+// ListQuarksSecrets fetches all the quarkssecrets from the namespace
+func ListQuarksSecrets(ctx context.Context, client crc.Client, namespace string) (*qsv1a1.QuarksSecretList, error) {
```
Comment (manno) [inline pkg/kube/util/resources/fetch.go:13]: The only advantage of this method, is that it's wrapping the error. Probably duplicating the error message..

**Review by manno — CHANGES_REQUESTED** [review]

File: pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go (hunk 4 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_reconciler.go:645]: True. I forgot to remove this part of the code from the PR.

File: pkg/kube/controllers/quarkssecret/quarkssecret_controller.go (hunk 7 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_controller.go:199]: yes. Added another check for the value too.

File: pkg/kube/util/resources/fetch.go (hunk 12 above)
Comment (rohitsakala) [reply pkg/kube/util/resources/fetch.go]: Removed whole GetReconciles package.

File: pkg/kube/util/reference/secret_references.go (hunk 11 above)
Comment (rohitsakala) [reply pkg/kube/util/reference/secret_references.go]: Removed whole GetReconciles package.

File: pkg/kube/util/reference/reconciles.go (hunk 10 above)
Comment (rohitsakala) [reply pkg/kube/util/reference/reconciles.go]: Removed whole GetReconciles package.

File: pkg/kube/util/reference/reconciles.go (hunk 9 above)
Comment (rohitsakala) [reply pkg/kube/util/reference/reconciles.go]: Removed whole GetReconciles package.

File: pkg/kube/util/reference/reconciles.go (hunk 8 above)
Comment (rohitsakala) [reply pkg/kube/util/reference/reconciles.go]: Removed whole GetReconciles package.

File: pkg/kube/controllers/quarkssecret/quarkssecret_controller.go (hunk 6 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_controller.go]: Changed the logic

File: pkg/kube/controllers/quarkssecret/quarkssecret_controller.go (hunk 5 above)
Comment (rohitsakala) [reply pkg/kube/controllers/quarkssecret/quarkssecret_controller.go]: Done

**Review by manno — APPROVED** [review]

**Pull request merged**